src/vmm/src/builder.rs

```diff
@@ -324,21 +278,123 @@ pub fn build_microvm(
         pio_device_manager,
     };
 
+    Ok((vmm, vcpus))
+}
+
+/// Builds and starts a microVM based on the provided MicrovmState.
+///
+/// An `Arc` reference of the built `Vmm` is also plugged in the `EventManager`, while another
+/// is returned.
+#[cfg(target_arch = "x86_64")]
+pub fn build_microvm_from_snapshot(
+    event_manager: &mut EventManager,
+    microvm_state: MicrovmState,
+    guest_memory: GuestMemoryMmap,
+    track_dirty_pages: bool,
+    seccomp_filter: BpfProgramRef,
+) -> std::result::Result<Arc<Mutex<Vmm>>, StartMicrovmError> {
+    use self::StartMicrovmError::*;
+    let vcpu_count =
+        u8::try_from(microvm_state.vcpu_states.len()).map_err(|_| InvalidMicrovmState)?;
+
+    // Build Vmm.
+    let (mut vmm, vcpus) = build_microvm_common(
+        event_manager,
+        guest_memory.clone(),
+        track_dirty_pages,
+        vcpu_count,
+        true,
+    )?;
+
+    // Restore kvm vm state.
+    vmm.vm.restore_state(&microvm_state.vm_state).unwrap();
+
+    // Restore devices states.
+    let mmio_ctor_args = MMIODevManagerConstructorArgs {
+        mem: guest_memory,
+        vm: vmm.vm.fd(),
+    };
+    vmm.mmio_device_manager =
+        MMIODeviceManager::restore(mmio_ctor_args, &microvm_state.device_states).unwrap();
+
+    // Move vcpus to their own threads and start their state machine in the 'Paused' state.
+    vmm.start_vcpus(vcpus, seccomp_filter).map_err(Internal)?;
+
+    // Restore vcpus kvm state.
+    vmm.restore_vcpu_states(microvm_state.vcpu_states).unwrap();
```

Comment:
> Do we need to explicitly call resume from the API to resume the VCPUs ? Can't we just do that here so a new API call is not necessary. IMO the only logical action after load is to resume.
> 

Hmm, it's a good point and would simplify things a bit. But since `CreateSnapshot` involves pausing and resuming to be done explicitly, we would lose I guess some kind of a symmetry this way :-??.
Maybe we should start thinking more about the config file option also for load, as you already proposed in another thread. We could add a new cmd line parameter, `--snapshot-file` or `--snapshot-config-file` which would include the load config and also the optional logger and metrics configs. 

> I guess this unwrap is not in the final version of the PR. Error needs to travel up the stack and presented to the user.

Reply:
> Do we need to explicitly call resume from the API to resume the VCPUs ? Can't we just do that here so a new API call is not necessary. IMO the only logical action after load is to resume.

Yes, that is the proposed and approved API flow.

We can change that outside this PR, as it implies getting buy-in from users and updating documentation as well.